File: backend/app/core/rbac.py
"""Backend role-based access control (Phase 3, §16).

Thin layer over :mod:`app.core.auth` that turns "membership was validated"
into role-specific authorization. Every dependency first requires a validated
membership (``require_organization_membership``) and then enforces the role.

Role hierarchy: viewer < researcher < admin < owner.

Only the FastAPI dependency functions below are meant to be used in routes:

    from fastapi import Depends
    from app.core.rbac import require_viewer, require_researcher

    @router.get("/{organization_id}/research")
    def run_research(
        membership=Depends(require_researcher()),
    ):
        ...
"""

from __future__ import annotations

from typing import Callable

from app.core.auth import (
    ROLE_RANK,
    Membership,
    require_organization_membership,
    require_role,
)

# Named canonical roles (shared single source of truth).
ROLE_OWNER = "owner"
ROLE_ADMIN = "admin"
ROLE_RESEARCHER = "researcher"
ROLE_VIEWER = "viewer"

# The hierarchy, most privileged first.
ROLE_HIERARCHY: tuple[str, ...] = (
    ROLE_OWNER,
    ROLE_ADMIN,
    ROLE_RESEARCHER,
    ROLE_VIEWER,
)

__all__ = [
    "ROLE_OWNER",
    "ROLE_ADMIN",
    "ROLE_RESEARCHER",
    "ROLE_VIEWER",
    "ROLE_HIERARCHY",
    "ROLE_RANK",
    "has_minimum_role",
    "require_viewer",
    "require_researcher",
    "require_admin",
    "require_owner",
    "require_organization_membership",
    "Membership",
]


def has_minimum_role(actual_role: str | None, required_role: str) -> bool:
    """Pure predicate: is ``actual_role`` at or above ``required_role``?"""
    if not actual_role:
        return False
    return ROLE_RANK.get(actual_role, 0) >= ROLE_RANK.get(required_role, 0)


# Dependencies that also enforce a minimum role on top of membership.
def require_viewer() -> Callable[..., Membership]:
    """Any member of the organization (viewer+)."""
    return require_role(ROLE_VIEWER)


def require_researcher() -> Callable[..., Membership]:
    """Researcher or above (can run research / contribute documents)."""
    return require_role(ROLE_RESEARCHER)


def require_admin() -> Callable[..., Membership]:
    """Admin or owner (can manage organization & members)."""
    return require_role(ROLE_ADMIN)


def require_owner() -> Callable[..., Membership]:
    """Only the owner (can delete the organization)."""
    return require_role(ROLE_OWNER)

File: frontend/lib/organizations/rbac.ts
/**
 * AgentFlow AI — Organization RBAC helpers (Phase 3, Part B).
 *
 * Small pure helpers that map a membership role onto the capabilities used
 * across the app (view / research / manage / delete).
 *
 * Role hierarchy: viewer < researcher < admin < owner.
 *
 * SECURITY NOTE: frontend RBAC is NOT the security boundary. These helpers
 * only drive what the UI shows / links to for a signed-in user whose
 * membership has already been validated against Supabase. PostgreSQL RLS and
 * the backend authorization layer (backend/app/core/auth.py + rbac.py)
 * remain authoritative and must be re-checked on every privileged action.
 */

import {
  ORGANIZATION_ROLES,
  ROLE_RANK,
  type OrganizationRole,
} from "@/lib/organizations/types";

/** True when `role` meets or exceeds the given `required` role. */
export function hasMinimumRole(
  role: OrganizationRole | string | null | undefined,
  required: OrganizationRole,
): boolean {
  if (!role) return false;
  const actualRank = ROLE_RANK[role as OrganizationRole];
  if (actualRank === undefined) return false;
  return actualRank >= ROLE_RANK[required];
}

/** A member may view the organization's content (any membership). */
export function canView(
  role: OrganizationRole | string | null | undefined,
): boolean {
  return hasMinimumRole(role, "viewer");
}

/** A member may run research / contribute documents (researcher+). */
export function canResearch(
  role: OrganizationRole | string | null | undefined,
): boolean {
  return hasMinimumRole(role, "researcher");
}

/** A member may manage the organization & members (admin+). */
export function canManageOrganization(
  role: OrganizationRole | string | null | undefined,
): boolean {
  return hasMinimumRole(role, "admin");
}

/** Only the owner may delete an organization. */
export function canDeleteOrganization(
  role: OrganizationRole | string | null | undefined,
): boolean {
  return role === "owner";
}

/** Roles at-or-above a given role, used for e.g. role pickers. */
export function rolesAtOrAbove(required: OrganizationRole): OrganizationRole[] {
  const min = ROLE_RANK[required];
  return ORGANIZATION_ROLES.filter((role) => ROLE_RANK[role] >= min);
}
